### PR TITLE
HTTP Client Implementation

### PR DESCRIPTION
**Pull request by briandowns**
Signed-off-by: Brian Downs <brian.downs@gmail.com>

### Well detailed description of the change :

Adds an HTTP client that is configurable and reusable.

Resolves: #

#537 

### Type of change:

- [ ] Bug fix

- [x] New feature

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping

- [ ] Tests have been updated to reflect the changes done within this PR (if applicable).

- [ ] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

### CODE REVIEW

**Comment by briandowns** [comment]
Tested with the following code:

```cs
import HTTP;
import JSON;
import System;

// main
{
    const opts = {
        "timeout": 20,
        "headers": [
            "Content-Type: application/json", 
            "Accept: application/json",
            "User-Agent: Dictu"
        ],
        "insecure": true,
        "keyFile": "",
        "certFile": "",
        "keyPasswd": ""
    };
    const ret = HTTP.newClient(opts);
    if (not ret.success()) {
        print(res.unwrapError());
        System.exit(1);
    }
    var httpClient = ret.unwrap();

    var res = httpClient.get("https://httpbin.org/get");
    if (not res.success()) {
        print(res.unwrapError());
        System.exit(1);
    }
    print("Status Code: {}".format(res.unwrap().statusCode));
    print(res.unwrap().json().unwrap());

    res = httpClient.get("https://httpbin.org/uuid");
    if (not res.success()) {
        print(res.unwrapError());
        System.exit(1);
    }
    //print(res.unwrap().statusCode);
    //print(res);
}
```

Results:

```
Status Code: 200
{"headers": {"Host": "httpbin.org", "Accept": "application/json", "Content-Type": "application/json", "User-Agent": "Dictu", "X-Amzn-Trace-Id": "Root=1-635f47b7-43af19f13ca0e7075eb70da9"}, "args": {}, "url": "https://httpbin.org/get", "origin": "24.251.231.19"}
Failed initialization
```

Current Status:

Working on figuring out why I'm getting the "failed to initialize" error when using the client again.

Was hopin to get your thoughts so far on this effort @Jason2605 before I go too much further.

**Comment by Jason2605** [comment]
I very much like the Dictu implementation you have there included as an example. Being honest, I'll need to brush up on my libcurl knowledge when it comes to reviewing the C code 😰. I'll also take a look to see if I can find anything about the failed init step on subsequent requests. Much appreciate the work gone into this too!

**Comment by briandowns** [comment]
Thanks! I'm gonna keep plugging away until it works.

**Comment by Jason2605** [comment]
Ahh I see, we're cleaning the httpClient on endRequest: https://github.com/dictu-lang/Dictu/blob/develop/src/optionals/http/http.c#L398-L400

**Comment by briandowns** [comment]
Ugh. I had a feeling it was in there but hadn't gotten deep enough. Thank you for the 👀! I'll fix this evening.

**Comment by Jason2605** [comment]
No worries! Seems I am getting a memory issue though, that will be a fun one to track down haha 😳

**Comment by briandowns** [comment]
I think I have all of the functionality and bugs worked out at this point. Marking this "ready for review".

File: src/optionals/http/http.c [hunk 1]
```diff
@@ -609,6 +613,428 @@ static Value post(DictuVM *vm, int argCount, Value *args) {
...
+
+    if (!IS_DICT(args[0])) {
+        runtimeError(vm, "Options dict passed to newClient() must be a dict.");
+        return EMPTY_VAL;
+    }
+
+    ObjDict *opts = AS_DICT(args[0]);
+
+    ObjAbstract *hc = newHttpClient(vm, opts);
+    return newResultSuccess(vm, OBJ_VAL(hc));
```
Comment (Jason2605) [inline src/optionals/http/http.c:1035]: I wonder, do we actually need to return a `Result<>` type here if we only ever return a successful result?

File: tests/http/client.du [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        
+        this.assertNil(ret);
+    }
+
+    testSetInsecure() {
+        const ret = this.httpClient.setInsecure(true);
+        this.assertNil(ret);
+
+        // const ret2 = this.httpClient.setInsecure(1);
+        // this.assertEmpty(ret2);
```
Comment (Jason2605) [inline tests/http/client.du:44]: I take it you meant to remove these?

**Review by Jason2605 — COMMENTED** [review]
Just a couple of really small comments, this is an awesome bit of work though, much needed improvements to the HTTP module!

File: src/optionals/http/http.c (hunk 1 above)
Comment (briandowns) [reply src/optionals/http/http.c]: Updated to not return the Result type. Tests, examples, and docs updated. Good call. That makes a lot more sense. Thanks!

**Comment by briandowns** [comment]
> Just a couple of really small comments, this is an awesome bit of work though, much needed improvements to the HTTP module!
Thanks! All set now I believe.

**Comment by Jason2605** [comment]
Thanks for this!

**Pull request merged**